src/dir.rs

```diff
@@ -0,0 +1,167 @@
+            }
+            Some(d) => Ok(Dir(d)),
+        }
+    }
+}
+
+// `Dir` is not `Sync`, but it's not reference-counted either, so it can be safely passed from one
+// thread to another.
+unsafe impl Send for Dir {}
+
+impl<'d> IntoIterator for &'d mut Dir {
+    type Item = Result<Entry>;
+    type IntoIter = Iter<'d>;
+    fn into_iter(self) -> Self::IntoIter {
+        Iter(self)
+    }
+}
+
+impl AsRawFd for Dir {
+    fn as_raw_fd(&self) -> RawFd {
+        unsafe { libc::dirfd(self.0.as_ptr()) }
+    }
+}
+
+impl fmt::Debug for Dir {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Dir")
+            .field("fd", &self.as_raw_fd())
+            .finish()
+    }
+}
+
+impl Drop for Dir {
+    fn drop(&mut self) {
+        unsafe { libc::closedir(self.0.as_ptr()) };
+    }
+}
+
+#[allow(missing_debug_implementations)]
+pub struct Iter<'d>(&'d mut Dir);
+
+impl<'d> Iterator for Iter<'d> {
+    type Item = Result<Entry>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        unsafe {
+            // Note: POSIX specifies that portable applications should dynamically allocate a
+            // buffer with room for a `d_name` field of size `pathconf(..., _PC_NAME_MAX)` plus 1
+            // for the NUL byte. It doesn't look like the std library does this; it just uses
+            // fixed-sized buffers (and libc's dirent seems to be sized so this is appropriate).
+            // Probably fine here too then.
+            let mut ent: Entry = Entry(::std::mem::uninitialized());
+            let mut result = ptr::null_mut();
+            if let Err(e) = Errno::result(readdir_r((self.0).0.as_ptr(), &mut ent.0, &mut result)) {
+                return Some(Err(e));
+            }
+            if result == ptr::null_mut() {
+                return None;
+            }
+            assert_eq!(result, &mut ent.0 as *mut dirent);
+            return Some(Ok(ent));
+        }
+    }
+}
+
+impl<'d> Drop for Iter<'d> {
+    fn drop(&mut self) {
+        unsafe { libc::rewinddir((self.0).0.as_ptr()) }
+    }
+}
+
+/// A directory entry, similar to `std::fs::DirEntry`.
+/// Note that unlike the std version, this may represent the `.` or `..` entries.
+#[derive(Copy, Clone)]
+pub struct Entry(dirent);
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Type {
+    Fifo,
+    CharacterDevice,
+    Directory,
+    BlockDevice,
+    File,
+    Symlink,
+    Socket,
+}
+
+impl Entry {
+    pub fn inode(&self) -> u64 {
+        self.0.d_ino
+    }
+    pub fn name(&self) -> &ffi::CStr {
+        unsafe { ::std::ffi::CStr::from_ptr(self.0.d_name.as_ptr()) }
+    }
+
+    /// Returns the type of this directory entry, if known.
+    /// See platform `readdir(3)` or `dirent(5)` manpage for when the file type is known;
+    /// notably, some Linux filesystems don't implement this. The caller should use `stat` or
+    /// `fstat` if this returns `None`.
+    pub fn type_(&self) -> Option<Type> {
```

Comment:
Since `Entry` is basically a replacement for `std::fs::DirEntry`, why not make the API as close as possible?  This method should be named `file_type`, and should return an object supporting methods like those of `std::fs::FileType`.

Reply:
Hmm. I'd like to preserve the ability to see exactly what the OS returned, for a couple reasons:

   * general educational value
   * you might want to use this to filter out uninteresting types without calling stat. If you decide to look more closely, you might want to do the stat in a different way (openat + fstat), and you might want to see the other fields without a second stat call.

My general impression was that nix is supposed to pretty closely match what the Unix systems offer, where the std library is more high-level. Does that sound right?

I made a similar decision to preserve the potentially-useful `.` and `..` entries (you might want to call `ino()` on them, for example) where std filters it out.

Reply:
That's reasonable.  But I would still rather call this method `file_type` or `filetype`.  The trailing underscore is weird and unprecedented.